File: tests/ane_load_test.py
#!/usr/bin/env python3
"""
ANE (Apple Neural Engine) Load Test for btop

This script generates load on the Apple Neural Engine to verify
that btop correctly displays ANE activity and power metrics.

Requirements:
- macOS with Apple Silicon (M1/M2/M3/M4)
- Python 3.8+
- coremltools: pip install coremltools
- numpy: pip install numpy

Usage:
    python3 ane_load_test.py [--duration SECONDS] [--intensity low|medium|high]
"""

import argparse
import time
import sys
import os

def check_apple_silicon():
    """Check if running on Apple Silicon."""
    import platform
    if platform.machine() != 'arm64':
        print("Error: This test requires Apple Silicon (M1/M2/M3/M4)")
        sys.exit(1)
    print(f"✓ Running on Apple Silicon ({platform.machine()})")

def create_test_model(size='medium'):
    """Create a simple CoreML model that runs on ANE."""
    try:
        import coremltools as ct
        import numpy as np
    except ImportError:
        print("Error: Required packages not installed.")
        print("Run: pip install coremltools numpy")
        sys.exit(1)

    # Model sizes for different intensity levels
    sizes = {
        'low': (64, 64, 32),
        'medium': (256, 256, 128),
        'high': (512, 512, 256)
    }

    input_size, hidden_size, output_size = sizes.get(size, sizes['medium'])

    print(f"Creating test model (intensity: {size})...")
    print(f"  Input: {input_size}, Hidden: {hidden_size}, Output: {output_size}")

    # Create a simple neural network using coremltools
    from coremltools.converters.mil import Builder as mb
    from coremltools.converters.mil.mil import types

    @mb.program(input_specs=[mb.TensorSpec(shape=(1, input_size), dtype=types.fp16)])
    def simple_nn(x):
        # First layer
        w1 = mb.const(val=np.random.randn(input_size, hidden_size).astype(np.float16) * 0.01)
        x = mb.matmul(x=x, y=w1)
        x = mb.relu(x=x)

        # Second layer
        w2 = mb.const(val=np.random.randn(hidden_size, hidden_size).astype(np.float16) * 0.01)
        x = mb.matmul(x=x, y=w2)
        x = mb.relu(x=x)

        # Third layer
        w3 = mb.const(val=np.random.randn(hidden_size, output_size).astype(np.float16) * 0.01)
        x = mb.matmul(x=x, y=w3)

        return x

    # Convert to CoreML model
    model = ct.convert(
        simple_nn,
        compute_units=ct.ComputeUnit.ALL,  # Allow ANE usage
        minimum_deployment_target=ct.target.macOS13
    )

    return model, input_size

def run_ane_load_test(duration=30, intensity='medium'):
    """Run continuous inference to generate ANE load."""
    import numpy as np

    check_apple_silicon()

    model, input_size = create_test_model(intensity)

    print(f"\n{'='*50}")
    print(f"ANE Load Test")
    print(f"{'='*50}")
    print(f"Duration: {duration} seconds")
    print(f"Intensity: {intensity}")
    print(f"\nMonitor ANE activity in btop (press '5' to show GPU panel)")
    print(f"{'='*50}\n")

    # Prepare input data
    input_data = {'x': np.random.randn(1, input_size).astype(np.float16)}

    start_time = time.time()
    iterations = 0
    last_report = start_time

    print("Running inference loop...")
    print("Time     | Iterations | Iter/sec")
    print("-" * 40)

    try:
        while time.time() - start_time < duration:
            # Run inference - this should use ANE
            _ = model.predict(input_data)
            iterations += 1

            # Report progress every second
            current_time = time.time()
            if current_time - last_report >= 1.0:
                elapsed = current_time - start_time
                rate = iterations / elapsed
                print(f"{elapsed:6.1f}s  | {iterations:10d} | {rate:8.1f}")
                last_report = current_time

    except KeyboardInterrupt:
        print("\n\nTest interrupted by user")

    elapsed = time.time() - start_time
    print("-" * 40)
    print(f"\nTest completed!")
    print(f"Total iterations: {iterations}")
    print(f"Total time: {elapsed:.1f} seconds")
    print(f"Average rate: {iterations/elapsed:.1f} iterations/second")
    print(f"\nCheck btop to verify ANE activity was recorded.")

def run_simple_test():
    """Run a simpler test using built-in macOS ML capabilities."""
    print("Running simple ANE test using Vision framework...")

    check_apple_silicon()

    # Use subprocess to run a simple Swift script that uses Vision/CoreML
    swift_code = '''
import Foundation
import CoreML
import Accelerate

// Simple matrix operations that may use ANE
let size = 1024
var matrix1 = [Float](repeating: 0, count: size * size)
var matrix2 = [Float](repeating: 0, count: size * size)
var result = [Float](repeating: 0, count: size * size)

// Initialize with random values
for i in 0..<(size * size) {
    matrix1[i] = Float.random(in: -1...1)
    matrix2[i] = Float.random(in: -1...1)
}

print("Starting matrix multiplication loop...")
let startTime = Date()
var iterations = 0

while Date().timeIntervalSince(startTime) < 30 {
    // Matrix multiply using Accelerate (may use ANE for large operations)
    vDSP_mmul(matrix1, 1, matrix2, 1, &result, 1, vDSP_Length(size), vDSP_Length(size), vDSP_Length(size))
    iterations += 1

    if iterations % 100 == 0 {
        let elapsed = Date().timeIntervalSince(startTime)
        print("Iterations: \\(iterations), Rate: \\(Double(iterations)/elapsed) iter/s")
    }
}

let elapsed = Date().timeIntervalSince(startTime)
print("\\nCompleted \\(iterations) iterations in \\(elapsed) seconds")
print("Rate: \\(Double(iterations)/elapsed) iterations/second")
'''

    import subprocess
    import tempfile

    with tempfile.NamedTemporaryFile(mode='w', suffix='.swift', delete=False) as f:
        f.write(swift_code)
        swift_file = f.name

    try:
        print("\nCompiling Swift test...")
        compile_result = subprocess.run(
            ['swiftc', '-O', '-o', '/tmp/ane_test', swift_file],
            capture_output=True, text=True
        )

        if compile_result.returncode != 0:
            print(f"Compilation failed: {compile_result.stderr}")
            return

        print("Running test (30 seconds)...")
        print("Monitor ANE activity in btop\n")

        subprocess.run(['/tmp/ane_test'])

    finally:
        os.unlink(swift_file)
        if os.path.exists('/tmp/ane_test'):
            os.unlink('/tmp/ane_test')

def main():
    parser = argparse.ArgumentParser(description='ANE Load Test for btop')
    parser.add_argument('--duration', type=int, default=30,
                        help='Test duration in seconds (default: 30)')
    parser.add_argument('--intensity', choices=['low', 'medium', 'high'],
                        default='medium', help='Test intensity (default: medium)')
    parser.add_argument('--simple', action='store_true',
                        help='Run simple test without coremltools dependency')

    args = parser.parse_args()

    print("=" * 50)
    print("ANE (Apple Neural Engine) Load Test for btop")
    print("=" * 50)
    print()

    if args.simple:
        run_simple_test()
    else:
        try:
            run_ane_load_test(args.duration, args.intensity)
        except ImportError:
            print("\ncoremltools not available, falling back to simple test...")
            run_simple_test()

if __name__ == '__main__':
    main()

File: tests/ane_load_test.swift
#!/usr/bin/env swift
/*
 * ANE (Apple Neural Engine) Load Test for btop
 *
 * This Swift script generates load on the Apple Neural Engine
 * to verify btop correctly displays ANE activity and power metrics.
 *
 * Compile: swiftc -O -o ane_load_test ane_load_test.swift
 * Run: ./ane_load_test [duration_seconds]
 */

import Foundation
import Accelerate
import CoreML

// MARK: - Simple MLModel for ANE testing

/// Create a simple computation that targets the ANE
func runANELoadTest(durationSeconds: Double = 30.0) {
    print("=" + String(repeating: "=", count: 49))
    print("ANE (Apple Neural Engine) Load Test for btop")
    print("=" + String(repeating: "=", count: 49))
    print()

    // Check for Apple Silicon
    #if !arch(arm64)
    print("Error: This test requires Apple Silicon (M1/M2/M3/M4)")
    exit(1)
    #endif

    print("✓ Running on Apple Silicon")
    print("Duration: \(Int(durationSeconds)) seconds")
    print()
    print("Monitor ANE activity in btop (press '5' to show GPU panel)")
    print("-" + String(repeating: "-", count: 49))
    print()

    // Use BNNS (Basic Neural Network Subroutines) which can utilize ANE
    let batchSize = 1
    let inputChannels = 256
    let outputChannels = 256
    let kernelSize = 3

    // Create input descriptor
    var inputDesc = BNNSNDArrayDescriptor(
        flags: BNNSNDArrayFlags(0),
        layout: BNNSDataLayoutImageCHW,
        size: (inputChannels, 64, 64, 0, 0, 0, 0, 0),
        stride: (0, 0, 0, 0, 0, 0, 0, 0),
        data: nil,
        data_type: .float,
        table_data: nil,
        table_data_type: .float,
        data_scale: 1.0,
        data_bias: 0.0
    )

    // Create output descriptor
    var outputDesc = BNNSNDArrayDescriptor(
        flags: BNNSNDArrayFlags(0),
        layout: BNNSDataLayoutImageCHW,
        size: (outputChannels, 64, 64, 0, 0, 0, 0, 0),
        stride: (0, 0, 0, 0, 0, 0, 0, 0),
        data: nil,
        data_type: .float,
        table_data: nil,
        table_data_type: .float,
        data_scale: 1.0,
        data_bias: 0.0
    )

    // Allocate memory
    let inputSize = inputChannels * 64 * 64
    let outputSize = outputChannels * 64 * 64
    var inputData = [Float](repeating: 0, count: inputSize)
    var outputData = [Float](repeating: 0, count: outputSize)

    // Initialize with random data
    for i in 0..<inputSize {
        inputData[i] = Float.random(in: -1...1)
    }

    print("Running neural network operations...")
    print()
    print("Time     | Operations | Ops/sec  | Status")
    print("-" + String(repeating: "-", count: 49))

    let startTime = Date()
    var operations = 0
    var lastReportTime = startTime

    // Main computation loop
    while Date().timeIntervalSince(startTime) < durationSeconds {
        // Perform matrix operations that may use ANE
        inputData.withUnsafeMutableBufferPointer { inputPtr in
            outputData.withUnsafeMutableBufferPointer { outputPtr in
                // Multiple matrix multiplications to generate load
                for _ in 0..<10 {
                    // Use vDSP for vectorized operations
                    var sum: Float = 0
                    vDSP_sve(inputPtr.baseAddress!, 1, &sum, vDSP_Length(inputSize))

                    // Matrix-vector multiplication simulation
                    vDSP_vsmul(inputPtr.baseAddress!, 1, &sum, outputPtr.baseAddress!, 1, vDSP_Length(min(inputSize, outputSize)))

                    // Additional operations
                    vDSP_vadd(inputPtr.baseAddress!, 1, outputPtr.baseAddress!, 1, outputPtr.baseAddress!, 1, vDSP_Length(min(inputSize, outputSize)))
                }
            }
        }

        operations += 10

        // Report progress every second
        let currentTime = Date()
        if currentTime.timeIntervalSince(lastReportTime) >= 1.0 {
            let elapsed = currentTime.timeIntervalSince(startTime)
            let rate = Double(operations) / elapsed
            print(String(format: "%6.1fs  | %10d | %8.1f | Running", elapsed, operations, rate))
            lastReportTime = currentTime
        }
    }

    let totalTime = Date().timeIntervalSince(startTime)
    print("-" + String(repeating: "-", count: 49))
    print()
    print("Test completed!")
    print("Total operations: \(operations)")
    print(String(format: "Total time: %.1f seconds", totalTime))
    print(String(format: "Average rate: %.1f operations/second", Double(operations) / totalTime))
    print()
    print("Check btop to verify ANE activity was recorded.")
}

/// Alternative test using large matrix operations
func runMatrixTest(durationSeconds: Double = 30.0) {
    print("Running large matrix multiplication test...")
    print()

    let matrixSize = 2048
    let elementCount = matrixSize * matrixSize

    var matrixA = [Float](repeating: 0, count: elementCount)
    var matrixB = [Float](repeating: 0, count: elementCount)
    var matrixC = [Float](repeating: 0, count: elementCount)

    // Initialize matrices
    for i in 0..<elementCount {
        matrixA[i] = Float.random(in: -1...1)
        matrixB[i] = Float.random(in: -1...1)
    }

    print("Matrix size: \(matrixSize) x \(matrixSize)")
    print("Starting multiplication loop...")
    print()

    let startTime = Date()
    var iterations = 0
    var lastReportTime = startTime

    while Date().timeIntervalSince(startTime) < durationSeconds {
        // Large matrix multiplication - uses Accelerate/ANE
        vDSP_mmul(matrixA, 1,
                  matrixB, 1,
                  &matrixC, 1,
                  vDSP_Length(matrixSize),
                  vDSP_Length(matrixSize),
                  vDSP_Length(matrixSize))

        iterations += 1

        let currentTime = Date()
        if currentTime.timeIntervalSince(lastReportTime) >= 1.0 {
            let elapsed = currentTime.timeIntervalSince(startTime)
            let rate = Double(iterations) / elapsed
            let gflops = (2.0 * Double(matrixSize * matrixSize * matrixSize) * rate) / 1e9
            print(String(format: "Time: %5.1fs | Iterations: %4d | Rate: %5.1f/s | ~%.1f GFLOPS", elapsed, iterations, rate, gflops))
            lastReportTime = currentTime
        }
    }

    let totalTime = Date().timeIntervalSince(startTime)
    let avgRate = Double(iterations) / totalTime
    let avgGflops = (2.0 * Double(matrixSize * matrixSize * matrixSize) * avgRate) / 1e9

    print()
    print("Test completed!")
    print("Total iterations: \(iterations)")
    print(String(format: "Average rate: %.1f iterations/second", avgRate))
    print(String(format: "Average performance: ~%.1f GFLOPS", avgGflops))
}

// MARK: - Main

let args = CommandLine.arguments
var duration: Double = 30.0
var useMatrixTest = false

for (index, arg) in args.enumerated() {
    if arg == "--matrix" || arg == "-m" {
        useMatrixTest = true
    } else if arg == "--duration" || arg == "-d", index + 1 < args.count {
        duration = Double(args[index + 1]) ?? 30.0
    } else if let d = Double(arg), index > 0 {
        duration = d
    } else if arg == "--help" || arg == "-h" {
        print("""
        ANE Load Test for btop

        Usage: ane_load_test [options] [duration]

        Options:
            --duration, -d <seconds>  Test duration (default: 30)
            --matrix, -m              Use matrix multiplication test
            --help, -h                Show this help

        Examples:
            ./ane_load_test              # Run for 30 seconds
            ./ane_load_test 60           # Run for 60 seconds
            ./ane_load_test --matrix     # Use matrix test
        """)
        exit(0)
    }
}

if useMatrixTest {
    runMatrixTest(durationSeconds: duration)
} else {
    runANELoadTest(durationSeconds: duration)
}
